### DISCUSSION TITLE
react-hook-form onsubmit data showing the defaultvalues, although watch prints updated value

### DISCUSSION THREAD

**Discussion by radikris** [react-hook-form/react-hook-form#9675] [Q&A] [answered]
Check question on stackoverflow: https://stackoverflow.com/questions/74901260/react-hook-form-onsubmit-data-showing-the-defaultvalues-although-watch-prints-u

*********************
I have a table which is responsive, but the layout is different so I have 2 different component for this.

This is a table, where the user can edit it's content.
Here is the layout:
[![enter image description here][1]][1]

When the user clicks on the green button, the editing will be done, and the values will be saved. It's a bit tricky, because the the button trigger the react-hook-form onSubmit which is in a different component solved with forwardRef.

    ("react-hook-form": "^7.38.0",)

Here is the EditableView code:

    import React, { forwardRef } from "react";
    import { useForm, useFieldArray } from "react-hook-form";
    import { useProfileContext } from "../../context/ProfileContext";
    import PrimaryButton from "../Button/PrimaryButton";
    
    const EditableDayMenu = forwardRef(({ weekOfYear, dayOfWeek, menu }, ref) => {
      const { register, control, handleSubmit, watch, getValues } = useForm({
        defaultValues: {
          menu: menu?.map((item) => {
            return {
              menuCategory: item.menuCategory,
              menuDescription: item.menuDescription,
            };
          }),
        },
      });
    
      const { fields, append, remove } = useFieldArray({
        control,
        name: "menu",
      });
      const { handleMenu } = useProfileContext();
    
      const onSubmit = (data) => {
        console.log(Data);
        handleMenu(weekOfYear, dayOfWeek, data);
      };
    
      console.log(watch());
    
      return (
        <form onSubmit={handleSubmit(onSubmit)}>
          {fields.map((item, index) => {
            return (
              <div className="" key={item.id}>
                <label className="block text-gray-700 text-sm font-bold mb-2">
                  Menü kategória
                </label>
                <input
                  className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
                  {...register(`menu.${index}.menuCategory`, { required: true })}
                />
                <label className="block text-gray-700 text-sm font-bold mb-2">
                  Menü leírása
                </label>
                <textarea
                  rows={5}
                  className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
                  {...register(`menu.${index}.menuDescription`, { required: true })}
                />
    
                <button
                  className="mb-3"
                  type="button"
                  onClick={() => remove(index)}
                >
                  Menü törlése
                </button>
              </div>
            );
          })}
          <button
            className="block lg:hidden xl:block mx-auto"
            type="button"
            onClick={() => {
              append({ menuCategory: "", menuDescription: "" });
            }}
          >
            <PrimaryButton text={"Új menü hozzáadása"} />
          </button>
          <button
            className="hidden lg:block xl:hidden mx-auto"
            type="button"
            onClick={() => {
              append({ menuCategory: "", menuDescription: "" });
            }}
          >
            <PrimaryButton text={"+"} />
          </button>
          <button hidden={true} ref={ref} type={"submit"} />
        </form>
      );
    });
    
    export default EditableDayMenu;
    EditableDayMenu.displayName = "EditableDayMenu";

The whole menu with the refs:

    
    const WeeklyMenu = ({ allMenu, isEditable = true }) => {
      const [currentWeek, setCurrentWeek] = useState();
      const [isEditView, setIsEditView] = useState(false);
      const fixWeek = Number(moment().format("w"));
      const arrLength = 7;
      const elRefs = React.useRef([]);
      const mobileelRefs = React.useRef([]);
      const [isMobileView, setIsMobileView] = useState();
      
      const handleResize = () => {
         window.innerWidth>1024 ? setIsMobileView(false) : 
         setIsMobileView(true)
      }

    
   
      if (elRefs.current.length !== arrLength) {
        // add or remove refs
        elRefs.current = Array(arrLength)
          .fill()
          .map((_, i) => elRefs.current?.[i] || createRef());
      }
      
      if (mobileelRefs.current.length !== arrLength) {
        // add or remove refs
        mobileelRefs.current = Array(arrLength)
          .fill()
          .map((_, i) => mobileelRefs.current?.[i] || createRef());
      }
    
     
      const menuThisWeek = menuOfTheWeek(currentWeek);
    
    
      return (
        <div className="text-Brand-Primary relative pt-4 lg:pt-0">
          {menuThisWeek != null && isEditable && (
            <div className="flex items-center gap-x-2 absolute right-0 w-full justify-center sm:w-fit">
              
              {!isEditView ? (
                <button
                  onClick={() => setIsEditView(true)}
                  className="bg-Brand-Secondary rounded-2xl px-3 py-2 text-white hover:bg-opacity-80"
                >
                  Menü szerkesztése
                </button>
              ) : (
                <button
                  onClick={() => {
                    setIsEditView(false);
                    const weekResponsiveButtons = isMobileView ? mobileelRefs : elRefs
                    console.log(isMobileView)
                    weekResponsiveButtons.current.forEach((ref) => {
                      console.log(ref.current);
                      ref.current?.click();
                    });
                  }}
                  className="bg-green-500 rounded-2xl px-3 py-2 text-white hover:bg-opacity-80"
                >
                  Mentés
                </button>
              )}
            </div>
          )}
          <div className="block lg:hidden h-12" />
          <p className="text-3xl">Heti menü</p>
         
          <div className="w-full space-x-3 hidden lg:flex">
          <DesktopWeeklyMenu
              isEditView={isEditView}
              weeks={getCurrentWeek(fixWeek - currentWeek)}
              menus={menuThisWeek}
              currentWeek={currentWeek}
              elRefs={elRefs}
              menuThisWeek={menuThisWeek}
            />
          </div>
          {(
            <MobileWeeklyMenu
              isEditView={isEditView}
              weeks={getCurrentWeek(fixWeek - currentWeek)}
              menus={menuThisWeek}
              currentWeek={currentWeek}
              elRefs={mobileelRefs}
              menuThisWeek={menuThisWeek}
            />
          )}
    
          {menuThisWeek == null && (
            <UnlockMenuWeek weekOfYear={currentWeek} isEditable={isEditable} />
          )}
        </div>
      );
    };
    
    export default WeeklyMenu;

When I try to edit in the desktop layout, it works perfectly, in the mobile layout it's not, although we are using the same EditableDayMenu component in both:

      const onSubmit = (data) => {
        handleMenu(weekOfYear, dayOfWeek, data)
    };
        
        console.log(watch())

When we try to submit the form, the data only shows the old (defaultValues), but the console.log(watch()) prints out the latest updated fields, so I don't understand why it is not the same in the onsubmit (and what is the difference between my desktop and mobile approach?)

Desktop layout:

    import React from 'react'
    import EditableDayMenu from './EditableDayMenu';
    
    const DesktopWeeklyMenu = ({
        isEditView,
        weeks,
        menus,
        currentWeek,
        elRefs,
        menuThisWeek,
    }) => {
      return (
        <table className="table-fixed w-[90%] mx-auto border-2 border-gray-500 rounded-xl">
        <thead>
          <tr>
            {weeks.map((item) => {
              return (
                <th
                  key={item}
                  className="border border-gray-400 text-Brand-Secondary px-1 xl:px-6"
                >
                  {item}
                </th>
              );
            })}
          </tr>
        </thead>
        <tbody>
          <tr>
            {menuThisWeek?.map((day, dayOfWeek) => {
              return (
                <td
                  key={day + dayOfWeek}
                  valign="top"
                  className="border border-gray-400 p-1 2xl:p-3"
                >
                  <div className="flex flex-col gap-y-3">
                    {!isEditView &&
                      day.day?.map((menu, idx) => {
                        return (
                          <div key={menu.menuCategory + idx}>
                            <p className="font-bold text-lg">
                              {menu.menuCategory}
                            </p>
                            <p className="font-base text-gray-700 break-all">
                              {menu.menuDescription}
                            </p>
                          </div>
                        );
                      })}
                  </div>
                  {isEditView && (
                    <EditableDayMenu
                      ref={elRefs.current[dayOfWeek]}
                      weekOfYear={currentWeek}
                      dayOfWeek={dayOfWeek}
                      menu={day.day}
                    />
                    
                  )}
                </td>
              );
            })}
          </tr>
        </tbody>
      </table>
        )
    }
    
    export default DesktopWeeklyMenu

Mobile layout:

    import React from "react";
    import EditableDayMenu from "./EditableDayMenu";
    import UnlockMenuWeek from "./UnlockMenuWeek";
    
    const MobileWeeklyMenu = ({
      isEditView,
      weeks,
      menus,
      currentWeek,
      elRefs,
      menuThisWeek,
    }) => {
    
      return (
        <table className="w-full mx-auto border-collapse border-spacing-10 block lg:hidden whitespace-normal border-2 border-gray-500 rounded-xl overflow-x-auto">
          {weeks.map((week, dayOfWeek) => {
            return (
              <tr key={week}>
                <th className="border w-[30%] border-gray-400 text-Brand-Secondary px-1 xl:px-6">
                  {week}
                </th>
                <td valign="top" className={`border border-gray-400 p-3 ${menuThisWeek==null ? 'bg-slate-200' :''}`}>
                  {menus?.[dayOfWeek]?.day?.map((day, index) => {
                    return !isEditView ? (
                      <div key={day.menuCategory + index}>
                        <p className="font-bold text-lg">{day.menuCategory}</p>
                        <p className="font-base text-gray-700 break-all">
                          {day.menuDescription}
                        </p>
                      </div>
                    ) : (
                      <EditableDayMenu
                        key={day.menuCategory + index+"edit"}
                        ref={elRefs.current[dayOfWeek]}
                        weekOfYear={currentWeek}
                        dayOfWeek={dayOfWeek}
                        menu={menus?.[dayOfWeek]?.day}
                      />
                    );
                  })}
                </td>
              </tr>
            );
          })}
         
        </table>
      );
    };
    
    export default MobileWeeklyMenu;

Here is an example of the log in the mobile layout:
[![enter image description here][2]][2]
`Watch`:

    {menu: Array(3)}menu: Array(3)0: {menuCategory: 'EDIT CATEGORY NAME', menuDescription: 'Leves minden leírva blabla, második'}1: ...

OnSubmit:

     {menu: Array(3)}menu: Array(3)0: {menuCategory: 'a1', menuDescription: 'Leves minden leírva blabla, második'}1: ...

After save:
[![enter image description here][3]][3]

You can see, that the edited value will not be stored, it's very strange, because the same works in the desktop version, hope you guys can help me figure it out.

  [1]: https://i.stack.imgur.com/Zkhbo.png
  [2]: https://i.stack.imgur.com/G8qZH.png
  [3]: https://i.stack.imgur.com/pYh9s.png

**Comment by radikris** [comment]
@bluebill1049 can you please check this out, if not enough the context, I will send you the source code and you could try out, but this issue gives me a headace since 1.5 months, and I can't figure it out why the console.log(watch()) shows the correct value, but the onSubmit is empty..

**Reply by Moshyfawn** [reply]
Please, provide a **minimal** reproduction sandbox instead. We can't debug everyone's business apps :P

**Reply by radikris** [reply]
@Moshyfawn thanks for your reply: Here you are: https://codesandbox.io/p/sandbox/sparkling-snowflake-9nbzej?file=%2FREADME.md


As I said the desktop editor working properly, but the responsive mobile not:
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/55052220/215320958-27017a45-c3cd-4867-892b-848c7388ba9b.png">

<img width="551" alt="image" src="https://user-images.githubusercontent.com/55052220/215320973-22fff700-25b6-4433-ba4d-6ddf17fccf89.png">

**Comment by leapful** [comment, marked as answer]
@radikris 

The root cause is because you are using CSS to show or hide desktop or mobile layout by using media query which leads to register same field name on both 2 layouts and causes wrong handling form values.

Switching to use `isMobileView` flag in order to render or not desktop and mobile layout will fix the problem.

Working demo: https://codesandbox.io/p/sandbox/tender-hypatia-2k63my?file=%2Fcomponents%2FWrapper.jsx

**Reply by radikris** [reply]
hey @leapful I checked your demo, and for me it is not working properly:
<img width="530" alt="image" src="https://user-images.githubusercontent.com/55052220/215705838-40f94bf0-cd07-4d2b-b5ed-cf3b2dba38c2.png">

I could add new things on an empty day but cannot edit the earlier one (like first thing is not editable)

**Reply by leapful** [reply]
@radikris 

Your implementation of MobileWeeklyMenu is wrong which render duplicated EditableDayMenu for each item of `day` of each week.

Just move EditableDayMenu outside the loop of `day` will solve the problem.

Working demo: https://codesandbox.io/p/sandbox/tender-hypatia-2k63my?file=%2Fcomponents%2FMobileWeeklyMenu.jsx

**Reply by radikris** [reply]
works like a charm, thank you very much @leapful !!!
